drivers/src/mailbox.rs

```diff
@@ -99,72 +128,42 @@ impl Mailbox {
     }
 }
 
-#[derive(Default)]
-/// Mailbox send protocol abstraction
-pub struct MailboxSendTxn {
-    /// Current state.
-    state: MailboxOpState,
+/// MailboxSendTxn protocol abstraction using typestate pattern.
+pub struct MailboxSendTxn<S> {
+    pub state: S,
+}
+impl Default for MailboxSendTxn<RdyForCmd> {
+    fn default() -> Self {
+        MailboxSendTxn { state: RdyForCmd }
+    }
 }
 
-impl MailboxSendTxn {
-    ///
-    /// Transitions from RdyCmd --> RdyForDlen
-    ///
-    pub fn write_cmd(&mut self, cmd: u32) -> CaliptraResult<()> {
-        if self.state != MailboxOpState::RdyForCmd {
-            raise_err!(InvalidStateErr)
-        }
+impl MailboxSendTxn<RdyForCmd> {
+    pub fn write_cmd(&mut self, cmd: u32) -> MailboxSendTxn<RdyForDlen> {
         let mbox = mbox::RegisterBlock::mbox_csr();
 
         // Write Command :
         mbox.cmd().write(|_| cmd);
-
-        self.state = MailboxOpState::RdyForDlen;
-        Ok(())
+        MailboxSendTxn { state: RdyForDlen }
     }
+}
 
-    ///
-    /// Writes number of bytes to data length register.
-    /// Transitions from RdyForDlen --> RdyForData
-    ///
-    pub fn write_dlen(&mut self, dlen: u32) -> CaliptraResult<()> {
-        if self.state != MailboxOpState::RdyForDlen {
-            raise_err!(InvalidStateErr)
-        }
-        let mbox = mbox::RegisterBlock::mbox_csr();
-
+impl MailboxSendTxn<RdyForDlen> {
+    /// Transition to the RdyForData state.
+    pub fn try_write_dlen(&mut self, dlen: u32) -> CaliptraResult<MailboxSendTxn<RdyForData>> {
         if dlen > MAX_MAILBOX_LEN {
             raise_err!(InvalidDlenErr);
         }
 
-        // Write Len in Bytes
-        mbox.dlen().write(|_| dlen);
-
-        self.state = MailboxOpState::RdyForData;
-        Ok(())
-    }
-
-    /// Transitions mailbox to RdyForData state and copies data to mailbox.
-    /// * 'cmd' - Command to Be Sent
-    /// * 'data' - Data Bufer
-    pub fn copy_request(&mut self, cmd: u32, data: &[u8]) -> CaliptraResult<()> {
-        if self.state != MailboxOpState::RdyForCmd {
-            raise_err!(InvalidStateErr)
-        }
-
-        self.write_cmd(cmd)?;
-
-        self.write_dlen(data.len() as u32)?;
-
-        // Copy data to mailbox
-        self.enqueue(data)?;
-
-        self.state = MailboxOpState::RdyForData;
+        let mbox = mbox::RegisterBlock::mbox_csr();
 
-        Ok(())
+        // Write Data Length :
+        mbox.dlen().write(|_| dlen);
+        Ok(MailboxSendTxn { state: RdyForData })
     }
-
-    fn enqueue(&self, buf: &[u8]) -> CaliptraResult<()> {
+}
+impl MailboxSendTxn<RdyForData> {
+    pub fn try_write_data(&self, buf: &[u8]) -> CaliptraResult<MailboxSendTxn<RdyForData>> {
         let remainder = buf.len() % size_of::<u32>();
```

Comment:
We could simplify the 4-byte-chunking logic with [`chunks_exact`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.chunks_exact).

Something like:
```rust
let chunks = buf.chunks_exact(size_of::<u32>());
let remainder = chunks.remainder();

for word in chunks {
    // write `word`
}

if !remainder.is_empty() {
    let last_word = remainder
        .iter()
        .rev()
        .fold(0, |w, &byte| (w << 8) | u32::from(byte));
    // write `last_word`
}
```

Reply:
Done. Thank you , Royce!